### ISSUE TITLE
Misplaced @test_case? 

### ISSUE THREAD

**Issue by xanderificnl** [enkessler/cuke_cataloger#6]
Thanks for this tool. I'm a bit new to Gherkin, so this may be my fault!

I've marked where I've condensed the feature file. 

```
Feature: User Registration

# removed 3 line user story.

  # <--- line 7
  Scenario: Create a basic user account, send a confirmation email and sign the user in.

    Given I register an account
    And I confirm the account
    Then I am signed in

  Rule: One account per email address
    Background:
      @test_case_10         # <-- line 16
      Given there is a user # <-- line 17 

    @wip
    @test_case_1
    Scenario: An account with the same email and password exists

# ... 8 more steps
```

I'm guessing `test_case_1` should be placed on line 7 and that `test_case_10` shouldn't be on line 16. Now the order of the test cases isn't really of any concern, but this error is: 

```
cuke_modeler-3.15.0/lib/cuke_modeler/parsing.rb:51:in `rescue in parse_text': Error encountered while parsing './features/user_registration.feature' (ArgumentError)
RuntimeError - (17:7): expected: #TagLine, #RuleLine, #Comment, #Empty, got 'Given there is a user
```

Fortunately, after manually moving `test_case_10` to line 7 the error is resolved.

**Comment by enkessler** [comment]
> Thanks for this tool. I'm a bit new to Gherkin, so this may be my fault!

You're welcome! I'm always glad when someone gets use out of these tools who isn't just me. And as long as the feature file that you started with is valid Gherkin, the tagged version that the cataloger spits out should also be valid. If it's not, then that's a problem that needs looking at. So thanks for opening the issue.

> I'm guessing test_case_1 should be placed on line 7 and that test_case_10 shouldn't be on line 16.

That is an odd placement of tags. I can't think of anything off hand that would cause it, so my only guess is that it's something to do with the presence of a `Rule`. If you don't mind sharing the feature file (or any feature file that will recreate the problem), I can track down the cause.

Come to think of it, I'm even more suspicious of the root cause being the rule nesting. I wrote this before rules were really a part of Gherkin, so feature files were entirely 'flat'. Because every addition of an ID tag will move the other tests down one line and because the tagging algorithm does them 'in order', there is an internal offset that is incremented each time when the cataloger has to figure out which line to place the new tag. Now that tests can be nested under other Gherkin objects (i.e. `Rule`s), that might change the order in which the tool is finding them and breaking the assumption that every test will come 'later' in the file than the previous test. The fact that the cataloger thinks that _that_ test is the 10th one but it is located near the top of the file would explain why it placed the tag about 9 lines off of where it should be. Especially if tests 1-9 are at the same level of nesting and tagged correctly.

I'll write up a test feature to check on my end but, still, having the original would help me know for sure.

**Comment by xanderificnl** [comment]
> You're welcome! I'm always glad when someone gets use out of these tools who isn't just me. And as long as the feature file that you started with is valid Gherkin, the tagged version that the cataloger spits out should also be valid. If it's not, then that's a problem that needs looking at. So thanks for opening the issue.

Thanks for the reply. I thought the feature file was valid because Cucumber didn't seem to have an issue with it. Searching for confirmation, I just installed `cuke_linter` and It tells a different story, or it shows just how new Cucumber/ Gherkin and BDD is for me. So my feature file may be causing this issue. In which case, I do apologize. 

For completeness, here's cuke_linter's output & below it my feature file.

```
❯ cuke_linter | grep -v '^[[:space:]]'
FeatureWithTooManyDifferentTagsLinter
TestWithNoActionStepLinter
TestWithNoVerificationStepLinter
StepWithTooManyCharactersLinter
ExampleWithoutNameLinter
SingleTestBackgroundLinter

15 issues found
```

For completeness, this is the feature file:

```
❯ sed 's/-/app_name/I; /@test_case_/d' ./features/user_registration.feature
Feature: User Registration

  As a visitor,
  I want to Sign up,
  so that I can use app_name

  Scenario: No account exists yet
  Create a basic user account, send a confirmation email and sign the user in.

    Given I register an account
    And I confirm the account
    Then I am signed in

  Rule: One account per email address
    Background:
      Given a confirmed user with the email "b.sisko@starfleet.ufp" and the password "ds9fever"

    @wip
    Scenario: An account with the same email and password exists

    Sign the user in.

      When I register an account
      Then I am signed in
      And I should see "You didn't have to sign up because your account already exists."

    Scenario: An account with the same email but a different password exists

    Send a forgotten password email.

      When the user above has the password "^foo^bar^"
      And I register an account
      Then I reset my password

    @oos
    Scenario: Changing the email address of one account to that of another initiates a merge request

    Benjamin has 2 accounts ben@sisko.org and captain@sisko.org. If he changes the email address of
    ben@sisko.org to captain@sisko.org the usual "please confirm your email address" mail is sent to
    captain@sisko.org. If Benjamin confirms the change, he'll be presented with the account merge
    procedure.

  Rule: An account requires at least one (valid) email address
    Background:
      # todo: refactor to use a table
      Given there is a user with the email "b.sisko@starfleet.ufp" and the password "ds9fever"

    Scenario Outline: Input validation fails
      When my "email" is "<email>"
      Then validation fails

      Examples:
        | email                | test_case_id |
        |                      | 12-1         |
        | .invalid@example.org | 12-2         |

  Rule: Accounts must be confirmed
    Background:
      Given there is a user with the email "b.sisko@starfleet.ufp" and the password "ds9fever"

    Scenario: No confirmation email is received because a typo in the address

    After creating their account they're signed in and are able to correct any typo(s)
    in their email address. As per policy, new email addresses are required to be
    confirmed as well.

    Scenario: New and unconfirmed accounts expire after 3 days

    New accounts that remain unconfirmed after 3 days are removed as they won't contain
    any data.

      # todo: implement clockwork and a background runner
      Given it is 3 days later
      When I confirm the account
      Then I should see "The confirmation code is invalid."


    Scenario: Changing an accounts email address requires a reconfirmation

    The accounts email address remains unchanged until it is confirmed.

    Scenario: Accounts are activated after following the link in the confirmation email
      Then I confirm the account

    Scenario: Accounts are activated after inputting the token in the confirmation form
      Then I confirm the account via the form

  Rule: 2-factor authentication is enforced with weak passwords

  Users of weak/unsafe passwords always receive a magic login link after signing in.
```

**Comment by enkessler** [comment]
>  I thought the feature file was valid because Cucumber didn't seem to have an issue with it. Searching for confirmation, I just installed `cuke_linter` and It tells a different story, or it shows just how new Cucumber/ Gherkin and BDD is for me. So my feature file may be causing this issue.

Your feature file is valid. `cuke_cataloger` and `cuke_linter` both use the same parser that `cucumber` does, under the covers. So if your feature were not syntactically valid then you would see a parsing error (like you do _after_ the cataloger adds the ID tags in wrong places) whenever you tried to use either of them. CukeLinter is more about catching stylistic concerns and enforcing good practices. So even if the linter thinks that the feature looks weird the feature should still run fine.

> For completeness, this is the feature file:

Thanks! This looks like it fits my hunch that test nesting is the root cause. I'll try to confirm and get a fix out this week. Also, good job on your taste in Start Trek series.

**Comment by enkessler** [comment]
@xanderificnl This should now be fixed with the `2.0.0` release. Feel free to check it out and confirm.

**Comment by xanderificnl** [comment]
> Also, good job on your taste in Start Trek series.

Haha, haven't found one I didn't like. They certainly spruce up testing, turns out Picard is amazing with a calculator! ;-)

> @xanderificnl This should now be fixed with the `2.0.0` release. Feel free to check it out and confirm.

Thank you! I can confirm it works: 

```
❯ grep '@test_case' user_registration.feature
  @test_case_1
    @test_case_2
    @test_case_3
    @test_case_4
    @test_case_5
    @test_case_6
    @test_case_7
    @test_case_8
    @test_case_9
    @test_case_10
```

Visual inspection:

![image](https://user-images.githubusercontent.com/8600870/152671061-d4d42305-35b9-4331-881b-921d0c22f764.png)

**Comment by enkessler** [comment]
Glad to hear it! Thanks again for bringing this to my attention.
